rmw_connext_cpp/src/functions.cpp

```diff
@@ -929,11 +931,78 @@ rmw_create_client(
     // error string was set within the function
     goto fail;
   }
+
   if (!get_datawriter_qos(participant, datawriter_qos)) {
     // error string was set within the function
     goto fail;
   }
 
+  switch (qos_profile.history) {
+    case RMW_QOS_POLICY_KEEP_LAST_HISTORY:
+      datawriter_qos.history.kind = DDS_KEEP_LAST_HISTORY_QOS;
+      datareader_qos.history.kind = DDS_KEEP_LAST_HISTORY_QOS;
+      break;
+    case RMW_QOS_POLICY_KEEP_ALL_HISTORY:
+      datawriter_qos.history.kind = DDS_KEEP_ALL_HISTORY_QOS;
+      datareader_qos.history.kind = DDS_KEEP_ALL_HISTORY_QOS;
+      break;
+    case RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT:
+      break;
+    default:
+      RMW_SET_ERROR_MSG("Unknown QoS history policy");
+      goto fail;
+  }
+
+  switch (qos_profile.reliability) {
+    case RMW_QOS_POLICY_BEST_EFFORT:
+      datawriter_qos.reliability.kind = DDS_BEST_EFFORT_RELIABILITY_QOS;
+      datareader_qos.reliability.kind = DDS_BEST_EFFORT_RELIABILITY_QOS;
+      break;
+    case RMW_QOS_POLICY_RELIABLE:
+      datawriter_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
+      datareader_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
+      break;
+    case RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT:
+      break;
+    default:
+      RMW_SET_ERROR_MSG("Unknown QoS reliability policy");
+      goto fail;
+  }
+
+  if (qos_profile.depth != RMW_QOS_POLICY_DEPTH_SYSTEM_DEFAULT) {
+    datareader_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+    datawriter_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+  }
+
+  // ensure the history depth is at least the requested queue size
+  assert(datawriter_qos.history.depth >= 0);
+  if (
+    datawriter_qos.history.kind == DDS::KEEP_LAST_HISTORY_QOS &&
+    static_cast<size_t>(datawriter_qos.history.depth) < qos_profile.depth
+  )
+  {
+    if (qos_profile.depth > (std::numeric_limits<DDS_Long>::max)()) {
+      RMW_SET_ERROR_MSG(
+        "failed to set history depth since the requested queue size exceeds the DDS type");
+      goto fail;
+    }
+    datawriter_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+  }
+  // ensure the history depth is at least the requested queue size
+  assert(datareader_qos.history.depth >= 0);
+  if (
+    datareader_qos.history.kind == DDS::KEEP_LAST_HISTORY_QOS &&
+    static_cast<size_t>(datareader_qos.history.depth) < qos_profile.depth
+  )
+  {
+    if (qos_profile.depth > (std::numeric_limits<DDS_Long>::max)()) {
+      RMW_SET_ERROR_MSG(
+        "failed to set history depth since the requested queue size exceeds the DDS type");
+      goto fail;
+    }
+    datareader_qos.history.depth = static_cast<DDS_Long>(qos_profile.depth);
+  }
+
```

Comment:
Wouldn't it make sense to move all this code into the `get_datareader_qos` / `get_datawriter_qos` functions where all the other qos properties are set? That would also avoid have the same ~60 lines in four locations.